### ISSUE TITLE
Silent integer overflow when using weights in histogram

### ISSUE THREAD

**Issue by wdecoster** [numpy/numpy#13879]
I use the `weights` argument in matplotlib.pyplot.hist() to create a weighted histogram. As far as I know, this passes on the calculations to numpy, so therefore I directly open the issue here. I noticed that the result of the `weights` argument depends on the dtype of the array. If I use an `int64` array things are as expected (see plot and code snippet below). However, if I use a `uint16` then the bin height is not the sum of the numbers in the bin. It looks a lot like the integer overflowed, and as such is not showing the sum for that bin.

In the plots below I combine a "normal" histogram showing just the frequency of the numbers in the array (top subplot), with a weighted histogram (bottom subplot)

### Reproducing code example:

**Works as expected:**
```
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
array = np.array([40017, 40324, 41151, 42109, 42897, 43388, 43756, 44380, 44569,  45578, 45589, 47422])
plt.subplot(2,1,1)
plt.xlim(40000,50000)
_ = plt.hist(array, bins=20, weights=None)
plt.ylim(0,5)
plt.subplot(2,1,2)
plt.xlim(40000,50000)
_ = plt.hist(array, bins=20, weights=array)
```

Specifically, looking at the bin close to x=46000 you'll see that there are 2 values in that bin in the upper subplot [45578, 45589]. As such their sum is ~91000, which seems okay in the bottom subplot.

![image](https://user-images.githubusercontent.com/11660522/60424982-d5eccb80-9bf1-11e9-8a64-e1df767f10b9.png)

**Shows weird behavior:**
```
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
array = np.array([40017, 40324, 41151, 42109, 42897, 43388, 43756, 44380, 44569,  45578, 45589, 47422], dtype='uint16')
plt.subplot(2,1,1)
plt.xlim(40000,50000)
_ = plt.hist(array, bins=20, weights=None)
plt.ylim(0,5)
plt.subplot(2,1,2)
plt.xlim(40000,50000)
_ = plt.hist(array, bins=20, weights=array)
```

Here again, looking at the bin close to x=46000 there are two values in the bin in the upper subplot, but in the bottom subplot the height is only ~35000 (based on eyeballing).

![image](https://user-images.githubusercontent.com/11660522/60425005-e604ab00-9bf1-11e9-8f4d-83544948aab7.png)

### Error message:

None raised

### Numpy/Python version information:

1.16.2 3.6.7 | packaged by conda-forge | (default, Feb 20 2019, 02:51:38) 
[GCC 7.3.0]

**Comment by wdecoster** [comment]
This seems to be expected behavior, but at least to me surprising nonetheless. I get a similar result when just doing a sum of two arrays:

```
a = np.array([50000], dtype='uint16')
b = np.array([40000], dtype='uint16')
a + b
> array([24464], dtype=uint16)
```

I guess I'll have to upcast my array.

**Comment by elarson425** [comment]
Hi @wdecoster I am experiencing the same issue you outlined above with NanoPlot v.1.43.1 - is it possible that this issue persists in later versions? See my example attached here. Any help would be greatly appreciated!
![newplot (1)](https://github.com/user-attachments/assets/a8410784-c300-4aa1-908b-77442e8d3e44)
![newplot](https://github.com/user-attachments/assets/17793c42-09eb-4851-85e3-84c2c3ee9a67)

**Comment by wdecoster** [comment]
I'm sorry, but what value exactly is unexpected? Is this an enormously large dataset?

**Comment by elarson425** [comment]
Hello! Thank you for getting back to me! I'm not sure of the exact values,
but I would expect it to be in the millions for sure. Here are summary
statistics from that same run. Over 11 billion total bases. It is a large
dataset.

[image: image.png]

**Comment by wdecoster** [comment]
I see your top graph has millions (M), do you then mean you don't see that in the bottom graph with the log transformation?

**Comment by elarson425** [comment]
Yes exactly!Sent from my iPhoneOn Jan 10, 2025, at 2:43 PM, Wouter De Coster ***@***.***> wrote:﻿
I see your top graph has millions (M), do you then mean you don't see that in the bottom graph with the log transformation?

**Comment by elarson425** [comment]
Hello! Just curious if you any more thoughts on what could be happening
here? Any insight would be truely appreciated!!

**Comment by wdecoster** [comment]
I don't know for sure there is a problem here. The different bin ranges in your top graph put many reads in the same bin, leading to a high number of reads there, while it is more spread out over multiple bins in the bottom graph. Does that make sense to you?
